core/src/main/java/org/htsjdk/core/utils/PathSpecifier.java

```diff
@@ -0,0 +1,238 @@
+    private String          pathFailureReason;  // cache the reason for "toPath" conversion failure
+
+    private transient Path cachedPath;         // cache the Path associated with this URI if its "Path-able"
+
+    public PathSpecifier(final String rawInputString) {
+        ParamUtils.nonNull(rawInputString);
+        this.rawInputString = rawInputString;
+
+        URI tempURI;
+        try {
+            // If the input URI already has a scheme (including a "file" scheme), we assume its already properly
+            // escaped. If no scheme component is present, then assume its a raw path on the local file system, so
+            // try to get a Path first, and then recreate the URI by retrieving it from the resulting Path. This
+            // ensures that input strings that contain embedded characters that would otherwise be interpreted as
+            // URI syntax (such as embedded fragment specifiers ("#") that are valid in file names) are properly
+            // escaped.
+            tempURI = new URI(rawInputString);
+            if (!tempURI.isAbsolute()) {
+                // NOTE: This case (no scheme) is the only one where we resolve the URI to a Path at construction time.
+                try {
+                    setCachedPath(Paths.get(rawInputString));
+                    tempURI = getCachedPath().toUri();
+                } catch (InvalidPathException e) {
+                    throw new IllegalArgumentException(e.getMessage(), e);
+                }
+            }
+            uri = tempURI;
+            if (!uri.isAbsolute()) {
+                // assert the invariant that every URI we create has a scheme, even if the raw input string does not
+                throw new HtsjdkException("URI has no scheme");
+            }
+        } catch (URISyntaxException e) {
+            final String errorMessage = String.format("%s must be a valid URI. '%s'/'%s'", rawInputString, e.getMessage(), e.getReason());
+            throw new IllegalArgumentException(errorMessage);
+        }
+    }
+
+    @Override
+    public boolean isNIO() {
+        // try to find a provider; assume that our URI always has a scheme
+        for (FileSystemProvider provider: FileSystemProvider.installedProviders()) {
+            if (provider.getScheme().equalsIgnoreCase(uri.getScheme())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public boolean isPath() {
+        try {
+            return getCachedPath() != null || toPath() != null;
+        } catch (ProviderNotFoundException |
+                FileSystemNotFoundException |
+                IllegalArgumentException |
+                HtsjdkException |
+                AssertionError e) {
+            // jimfs throws an AssertionError that wraps a URISyntaxException when trying to create path where
+            // the scheme-specific part is missing or incorrect
+            pathFailureReason = e.getMessage();
```

Comment:
If the message is empty or uninformative, the failure reason is not really useful; sometimes the useful message comes from the stacktrace. I would say to at least test if the message is empty, and if it is at least specify the class for the exception.

Reply:
Actually, I think that it will be much better to store the exception to let the user handle it by themselves as they wish.